src/main/java/io/stargate/sgv2/jsonapi/service/operation/reranking/RerankingMetrics.java

```diff
@@ -0,0 +1,241 @@
+    Objects.requireNonNull(sample, "Timer.Sample cannot be null");
+
+    // --- Tenant-Specific Timer ---
+    // Build tags for the tenant timer
+    Tags tenantTags =
+        new RerankingTagsBuilder()
+            .withTenant(requestContext.getTenantId().orElse(UNKNOWN_VALUE))
+            .withKeyspace(schemaObject.name().keyspace())
+            .withTable(schemaObject.name().table())
+            .build();
+    // Get the tenant timer instance
+    Timer tenantTimer = meterRegistry.timer(RERANK_TENANT_CALL_DURATION_METRIC, tenantTags);
+    // Stop the sample against the tenant timer. This records the duration AND returns it.
+    long durationNanos = sample.stop(tenantTimer);
+
+    // --- Overall Timer ---
+    // Build tags for the overall timer
+    Tags allTags =
+        new RerankingTagsBuilder()
+            .withProvider(classSimpleName(rerankingProvider.getClass()))
+            .withModel(rerankingProvider.modelName())
+            .build();
+    // Get the overall timer instance
+    Timer allTimer = meterRegistry.timer(RERANK_ALL_CALL_DURATION_METRIC, allTags);
+    // Manually record the exact same duration (obtained above) to the overall timer.
+    allTimer.record(durationNanos, TimeUnit.NANOSECONDS);
+  }
+
+  // --- Static Inner Tag Builder Class ---
+
+  /**
+   * Builder for creating {@link Tags} specific to reranking metrics.
+   *
+   * <p>Allows flexible combination of common reranking-related tags derived from the context
+   * provided to the outer {@link RerankingMetrics} instance.
+   */
+  public static class RerankingTagsBuilder {
+    // Use a Map to store tags and check for duplicates
+    private final Map<String, String> tagsMap;
+
+    public RerankingTagsBuilder() {
+      this.tagsMap = new HashMap<>();
+    }
+
+    public RerankingTagsBuilder withTenant(String tenantId) {
+      putOrThrow(TENANT_TAG, tenantId);
+      return this;
+    }
+
+    public RerankingTagsBuilder withKeyspace(String keyspace) {
+      putOrThrow(KEYSPACE_TAG, keyspace);
+      return this;
+    }
+
+    public RerankingTagsBuilder withTable(String table) {
+      putOrThrow(TABLE_TAG, table);
+      return this;
+    }
+
+    public RerankingTagsBuilder withProvider(String provider) {
+      // TODO: It's not easy to get the provider name as it is described in the config. So we use
+      // the class name to indicate the provider. Need to replace the class name in the future.
+      putOrThrow(RERANKING_PROVIDER_TAG, provider);
+      return this;
+    }
+
+    public RerankingTagsBuilder withModel(String modelName) {
+      putOrThrow(RERANKING_MODEL_TAG, modelName);
+      return this;
+    }
+
+    /** Helper method to check and put, throwing an exception on duplicate key */
+    private void putOrThrow(String key, String value) {
+      // If the key already exists, that means the caller is trying to set the same tag multiple
+      // times. Throw exception.
+      if (tagsMap.containsKey(key)) {
```

Comment:
Instead of checking separately, you can just check return value of `tagsMap.put()` -- it returns old value, if any. So something like:

    if (tagsMap.put(key, value) != null) { 
        throw new IllegalStateException(...);
    }